### PR TITLE
Demo v2: Auto-correct JSON format via loop

### PR DESCRIPTION
**Pull request by tholor**
## Motivation
Pipelines in Haystack 2.0 are **more flexible** and **more robust** than ever.
They allow you now to run loops within your pipeline. Let's look at why this matters.

## Exemplary Use Case: Validate & auto-correct the output format of an LLM
Let's imagine you want to extract structured information from a document in JSON format. 
As your further processing steps depend on the correctness of this JSON format, you want to make sure that the LLM really produces it in a robust way.

## Approach
We validate the output of the LLM against a pydantic schema. If the output is valid, we return it to the user. If not, we create another prompt that leverages the generated output + error message. To avoid infinite loops we limit it to max. 5 iterations.

parser_loop

Demo: [Colab](https://colab.research.google.com/drive/1_qMroflDWlqdQaoDPP-k8-9CutNcsvCp?usp=sharing) 

To Dos: 
- [x] We need optional variables in PromptTemplates (we are using a haystack branch so far)
- [x] Allow three variables in the prompt template so that we can include the error message (so far only the wrongly generated output is used)

### CODE REVIEW

File: auto_fixing_parser_v2/auto_fix_parser.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+            f.write(replies[0].replace("\n", "") + "\n")
+
+        # create a corrupt json with 40% probability (for demo purposes)
+        if random.randint(0, 100) < 10:
+            replies[0] = "Corrupt Key" + replies[0]
+        try:
+            output_dict = json.loads(replies[0])
+            self.pydantic_model.parse_obj(output_dict)
+            logging.info(f"Valid LLM output: {replies[0]}")
+            return {"valid": replies, "invalid": None, "error_message": None}
```
Comment (ZanSara) [inline auto_fixing_parser_v2/auto_fix_parser.py:55]:
```suggestion
            return {"valid": replies}
```

File: auto_fixing_parser_v2/auto_fix_parser.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        except (ValueError, ValidationError) as e:
+            logging.info(f"Invalid LLM output: {replies[0]}, error: {e}")
+            return {"valid": None, "invalid": replies, "error_message": str(e)}
```
Comment (ZanSara) [inline auto_fixing_parser_v2/auto_fix_parser.py:58]:
```suggestion
            return {"invalid": replies, "error_message": str(e)}
```

**Review by ZanSara — COMMENTED** [review]
Left two small fixes that make this work on Canals 0.10

**Comment by bilgeyucel** [comment]
Closing this one as we have the example as a tutorial: https://github.com/deepset-ai/haystack-tutorials/blob/main/tutorials/28_Structured_Output_With_Loop.ipynb

**Pull request closed without merging**